Feat: Support Compose Multiplatform

b-re-w opened the pull request:
안녕하세요!
CMP로 최근에 앱 개발을 하다가 Custom Slider를 구현할 일이 있어 검색을 통해 이 레포지토리를 알게되었습니다.
그런데 프로젝트에 implement 해보니 이 라이브러리에는 android target만 존재하고 있어서 그래들 스크립트에서 오류가 나더라고요. 그래서 JVM Desktop과 iOS 타겟도 추가하여 프로젝트 구조를 수정했습니다.

테스트 결과 전부 Pure Compose 코드로 작성되어서 그런지 작성된 코드를 별도로 수정하는 것 없이 바로 다른 플랫폼 타겟으로도 실행이 가능했습니다.
수정했는데 그냥 두기는 아까워서 풀리퀘 보내드립니다.
검토 및 수정 후 반영해주시면 감사하겠습니다!

아래는 iOS와 Desktop(Windows) 타겟 앱 실행 스크린샷입니다.

![iOS](https://github.com/seyoungcho2/FilledSliderCompose/assets/81044362/bb1aefef-3cfd-4e49-926f-49961789e243)
![Windows Desktop](https://github.com/seyoungcho2/FilledSliderCompose/assets/81044362/ccd0a602-1503-481f-9c66-b79ddefe4267)

감사합니다!!

P.S. Desktop 타겟으로는 common 모듈에서 implementation으로 filled-slider-compose 모듈을 불러오니 SliderColor 클래스를 런타임에 찾지 못하는 문제가 있어서 api로 불러오도록 해야 하는 것으로 보입니다.

seyoungcho2 commented:
@IRACK000 님 안녕하세요. 개발자 조세영입니다.
먼저 FilledSlider Compose에 Compose Multiplatform을 적용시킨 귀중한 Pull Request를 날려주셔서 감사하다는 말씀 드립니다.
다만 제가 아직 Compose Multiplatform을 사용해보지 못해 Compose Multiplatform Mobile에 관한 PR을 리뷰 할 수가 없습니다. 제가 현재 하고 있는 일 중 우선 순위가 높은 일이 있어 Compose Multiplatform을 공부하고 이 PR을 머지할 수 있는 건 내년 초에 가능할 것 같습니다. Compose Multiplatform을 리뷰할 만한 배경 지식이 쌓이는대로 리뷰하도록 하겠습니다. 

현재로서 제가 리뷰를 드릴 수 내용은 다음과 같습니다.
1. Pull Request의 단위가 분할 되어야 할 것 같습니다. 현재 올려주신 Pull Request에는 Gradle 업데이트, 모듈명 변경(sample-app --> android, main --> commonMain), target 버전 업데이트, 버전 관리 시 toml 적용 등이 한 번에 들어가 있는데, 한 번에 머지되어야 하는 코드 크기가 너무 커서 머지되면 영향 범위가 너무 넓습니다.
*만약 compose multiplatform을 지원하기 위해 target 버전 업데이트나 toml 적용이 필요가 없다면, Pull Request에서는 빼는 것이 좋을 것 같습니다. 만약 필요하다면, 사용하는 버전의 업데이트에 대한 PR이 먼저 올라가고, Compose Multiplatform에 대한 PR이 올라가야 할 것 같습니다.
2. Compose Multiplatform을 위해 모듈명을 바꿔야 한다면, sample-app --> android로 바꾸기 보다, 더욱 명확하게 sample-app-android, sample-app-desktop 로 바꾸는 것이 좋을 것 같습니다. 또한 모듈명에 common이 들어가는 경우 공통으로 사용되지만 어떤 경우에 사용되어야 하는지 모호한 경우가 많기 때문에 기능이 들어간 main 모듈은 그대로 이름을 두는 것이 좋을 것 같습니다.
3. 샘플 앱에서 ComposeSliderTheme 같이 함수의 선언부가 변경된 경우(Default 값이 없어진 경우)가 보이는데, 선언이 변경된 이유를 설명해주시면 좋을 것 같습니다.(프로덕션 코드가 아닌 샘플 코드라 중요한 부분은 아닙니다.)
4. 1번과 같은 이야기 인데, Source Directory(java to kotlin)가 바뀌어야 하는 것이 PR에 필수적이지 않은 경우 별도의 PR로 만들어지면 좋을 것 같습니다.
5. settings.gradle.kts에 jitpack.io는 Deprecated된 Repository여서 해당 부분 필수적이지 않다면 제거하는 것이 좋을 것 같습니다. 다른 repository에 대한 것도 Compose Multiplatform을 추가하는데 필수적이지 않은 경우 제거하는 것이 좋을 것 같습니다.  
6. minSDK가 26에서 24로 바뀌었는데, 일반적으로 처음 minSDK로 정한 값보다 낮은 값으로 내리는 경우는 불가피한 경우가 아니면 권장되지 않습니다. 특히 안드로이드의 경우 minSDK 26부터 쓸 수 있는 API들이 많고, Compose를 사용하는 애플리케이션은 주로 minSDK가 26이상인 경우가 많기 때문에 오픈소스 개발 시에는 minSDK를 26으로 유지하는 것이 좋을 것으로 보입니다. 

다시 한 번 라이브러리에 관심을 가져주시고, 많은 시간을 써주셔서 Compose Multiplatform을 적용시킨 Pull Request를 공유해 주셔서 감사합니다.

감사합니다. 좋은 저녁 되세요!
조세영 드림.

b-re-w committed:
chore: Upgrade Kotlin Version to 1.9.0

b-re-w committed:
Merge pull request #1 from IRACK000/chore/kotlin-version-1.9.0

chore: Upgrade Kotlin Version to 1.9.0

b-re-w committed:
chore: Set to use LibraryVersions

b-re-w committed:
Merge pull request #2 from IRACK000/chore/kotlin-version-1.9.0

chore: Set to use LibraryVersions

b-re-w committed:
feat: Suppoprt Multiplatform

b-re-w committed:
chore: Update Gradle Setting for MPP

b-re-w committed:
chore: Change srcDir Setting

b-re-w committed:
Merge pull request #4 from IRACK000/chore/kotlin-version-1.9.0

Chore/kotlin version 1.9.0

b-re-w committed:
Merge pull request #5 from IRACK000/feat/multiplatform

Feat/multiplatform

b-re-w committed:
fix: Roll-back to official-source sets

commonMain을 main으로, commonTest를 test로, androidUnitTest를 androidTest로 src 명을 변경하여 원본 소스 폴더 구조에서의 변경을 최소화 하려 했으나, android 타겟 외의 다른 타겟에서 정상적으로 빌드가 안되어서 롤백

b-re-w committed:
chore: Add iOS Support Setting

b-re-w committed:
chore: Set to use JetBrains Compose Gradle Plugin

기존에 사용하던 composeBom을 계속 사용하려고 해봤으나, composeBom은 android 타겟 외의 다른 타켓의 라이브러리를 불러와주지 않는 것으로 보여 어쩔 수 없이 1.5.0버전의 플러그인을 추가하였습니다. (사용중이던 Bom 버전이 불러오는 1.4.2버전은 kotlin 1.9.0과 호환되지 않아 버전을 올릴 수 밖에 없었습니다.)

b-re-w committed:
chore: Fix ClassCastException while Build

Upgrade Compose Compiler Version from 1.5.0 to 1.5.3

@Ref: https://github.com/JetBrains/compose-multiplatform/issues/3531

b-re-w committed:
Update .gitignore

b-re-w committed:
chore: Set Android Build Target for Multiplatform

b-re-w committed:
chore: Init sample-app module for multiplatform

b-re-w committed:
refactor: Move Sample Codes from target/android to common

b-re-w committed:
feat: Add Target-Specified Theme Code

b-re-w committed:
feat: Shared App View

b-re-w committed:
feat: Remove Nullable from Type.kt

b-re-w committed:
chore: Init Desktop Build Target for Multiplatform

b-re-w committed:
chore: Init iOS Build Target for Multiplatform

b-re-w committed:
refactor: Rename Source Module from java to kotlin in order to fix link error while building for desktop target

소스 디렉토리를 java로 유지하려 했으나 desktop 타겟 빌드시 코드를 찾지 못하는 문제가 발생하여 kotlin으로 이름 변경

b-re-w committed:
Merge pull request #7 from IRACK000/feat/multiplatform

fix: fix code link error (merge form feat/multiplatform)

b-re-w committed:
chore: Fix iOS Framework Setting

b-re-w committed:
chore: Fix iOS Framework Setting (module name)

b-re-w committed:
chore: Fix iOS Framework Setting (module path)

b-re-w committed:
fix: Add StatusBar Padding

b-re-w committed:
Merge pull request #8 from IRACK000/test/multiplatform-sample-apps

Test/multiplatform sample apps

IRACK000 commented:
안녕하세요! 리뷰해주신 대로 수정하느라 시간이 좀 걸려서 답변이 늦었습니다.


아래는 리뷰주신 내용에 대한 반영내역입니다.
- 일단 전반적으로 제가 라이브러리를 멀티플랫폼화 시키면서 기존의 빌드 구조를 바탕으로 작업하기 보다는 Compose Multiplatform 공식 템플릿에 기존 코드를 옮기는 방향으로 작업했기 때문에 변경의 영향이 너무 넓었던 것 같습니다. 제가 PR 받는 쪽 입장을 제대로 고려를 못했던 부분인것 같고, 그래서 변경 내역을 전부 지우고 다시 기존 프로젝트 구조를 기반으로 변경 내역을 하나씩 도입하였습니다. PR도 최대한 나눠보았습니다.


1. Pull Request의 단위가 분할 되어야 할 것 같습니다. 현재 올려주신 Pull Request에는 Gradle 업데이트, 모듈명 변경(sample-app --> android, main --> commonMain), target 버전 업데이트, 버전 관리 시 toml 적용 등이 한 번에 들어가 있는데, 한 번에 머지되어야 하는 코드 크기가 너무 커서 머지되면 영향 범위가 너무 넓습니다.
*만약 compose multiplatform을 지원하기 위해 target 버전 업데이트나 toml 적용이 필요가 없다면, Pull Request에서는 빼는 것이 좋을 것 같습니다. 만약 필요하다면, 사용하는 버전의 업데이트에 대한 PR이 먼저 올라가고, Compose Multiplatform에 대한 PR이 올라가야 할 것 같습니다.

1A. 말씀하신 부분 반영하기 위해서 독립적으로 실행 가능한 변경사항들끼리 최대한 분리하여 PR 만들었습니다. 모듈명과 프로젝트 구조도 기존 구조에서 변경하지 않기 위해서 최대한 노력했습니다. 그리고 필수가 아니었던 버전 관리 toml도 애초에 프로젝트 템플릿을 사용하지 않았기 때문에 굳이 적용하지 않았습니다. 타겟 버전 업데이트의 경우 멀티플랫폼을 지원하기 위해 코틀린 버전을 1.9.0으로 업데이트하고, 그에 따른 컴포즈 버전도 1.5.0으로 업데이트를 할 수밖에 없었습니다...(버전 안올리니까 오류가 나서 어쩔 수 없었습니다.) 버전 올리는 것에 대한 영향 검토해주시면 될거같습니다. #2 

2. Compose Multiplatform을 위해 모듈명을 바꿔야 한다면, sample-app --> android로 바꾸기 보다, 더욱 명확하게 sample-app-android, sample-app-desktop 로 바꾸는 것이 좋을 것 같습니다. 또한 모듈명에 common이 들어가는 경우 공통으로 사용되지만 어떤 경우에 사용되어야 하는지 모호한 경우가 많기 때문에 기능이 들어간 main 모듈은 그대로 이름을 두는 것이 좋을 것 같습니다.

2A. 타겟별 sample app 모듈명을 바꾸지 않으려 노력을 좀 많이 했습니다. 제가 원래 올렸던 PR의 프로젝트 구조는 컴포즈 멀티플랫폼 기본 템플릿이라 그걸 그대로 사용하면 되었었는데, 타겟별로 분리하지 않고 한 폴더 안에 다 모아두려 하니 변경해야 할 부분들이 많았습니다. 샘플 앱 관련 PR은 전부 하나의 PR #4에 모아두었습니다. 루트에 있던 common 모듈의 경우 CMP에서는 일반적으로 common에다 플랫폼 독립적인 코드를 작성하고, 각 타겟별 모듈(android, desktop, ios)에 플랫폼 종속적인 코드를 작성하는 방식으로 하다보니 common이 따로 나와있었는데, 아무래도 이 레포가 앱이 아닌 라이브러리에 대한 레포다 보니, 레포 루트에 android, desktop, ios, common이 다 나와있으면 정리 안된 느낌이 들거 같긴 해서 기존 sample-app폴더 안에 전부 정리해서 넣어두었습니다. common을 없애기 위해 sample-app 폴더가 common 역할을 하도록 했고, 타겟별 모듈은 sample-app 폴더 안에 target 폴더를 두어 그 안에 배치해두었습니다.
main모듈 이름을 그대로 두는 것이 좋겠다고 말씀하셔서 그렇게 해보려고 시도를 해봤습니다. main -> commonMain 이런 형태로 바꾼 것은 제가 그렇게 한게 아니라 멀티플랫폼 플러그인이 그렇게 되어 있어서 그런거였는데, source directory 폴더를 기존 안드로이드 모듈명으로 지정;kotlin.srcDirs("src/main")하여 빌드를 해보니, 안드로이드 타겟 빌드에서만 정상 작동하고, 나머지 타겟에서는 소스 폴더를 인식하지 못해서 다시 되돌려놓았습니다. [https://github.com/IRACK000/FilledSliderCompose/commit/068aaacbcc23dd0dbf19b81161a3287ce34c692c](url) 커밋 참고하시면 되겠습니다.
* [https://kmp.jetbrains.com/](url) 여기를 참고하시면, fleet 에디터를 기반으로 한 새로운 멀티플랫폼용 템플릿 프로젝트 생성기를 확인하실 수 있으며, 이 프로젝트 구조도 common을 사용하지 않으며 하나의 모듈에서 모든 타겟을 (ios를 제외한) 대응하는 것으로 보입니다. 다만 제가 아직 여기에 익숙하지가 않고, 타겟중 ios만 빠져나와 있는데다 fleet을 사용해야 원활하게 사용할 수 있는 듯 하여 사용하지 않았습니다. 나중에 한번 참고하시면 될거같습니다.

3. 샘플 앱에서 ComposeSliderTheme 같이 함수의 선언부가 변경된 경우(Default 값이 없어진 경우)가 보이는데, 선언이 변경된 이유를 설명해주시면 좋을 것 같습니다.(프로덕션 코드가 아닌 샘플 코드라 중요한 부분은 아닙니다.)

3A. 이 부분은 코드를 다 안보셔서 그런것 같습니다. ComposeSliderTheme의 경우 Platform 종속적인 코드가 포함되어 있습니다. 안드로이드 12부터 지원되는 dynamicColor를 지원하기 위해 안드로이드 버전을 체크하기 위한 Build클래스와 최종 결정된 컬러를 시스펨바에 반영하기 위해 호출되는 LocalView.current가 다른 플랫폼에서는 지원되지 않기 때문에 commonMain 내에서는 컴파일 될 수 없습니다. 따라서 actual과 expect 키워드를 사용해서 플랫폼별 코드를 분리해서 서로 다르게 ComposeSliderTheme를 컴파일 하도록 작성해두었었는데, default 값은 expect function에서 선언 가능하고 actual쪽에서는 사용하지 않기 때문에 빠져있는 것으로, 아무래도 androidMain쪽 코드만 보셨던 것 같습니다. 그래도 ComposeSliderTheme 자체가 expect-actual로 선언되는 것보다는 그 내부 코드들이 분리되는 편이 좋을 것 같아 새 PR에서는 컬러 선택부와 글꼴 선택부만 따로 분리하였습니다. [https://github.com/IRACK000/FilledSliderCompose/commit/7de8edce1cb5ac6f950c8e1ddd61484d4aee713d](url)

4. 1번과 같은 이야기 인데, Source Directory(java to kotlin)가 바뀌어야 하는 것이 PR에 필수적이지 않은 경우 별도의 PR로 만들어지면 좋을 것 같습니다.

4A. java to kotlin으로의 이름 변경이 저도 처음에는 필수가 아니라고 생각하고 있었어서 java로 그대로 놔둬보려고 시도해봤습니다. 테스트 결과 안드로이드 타겟 빌드시에는 모든 코드가 java 폴더 아래에 위치하고 있어도 빌드가 되지만, desktop과 ios 빌드시에는 kotlin 폴더 아래에 위치하고 있어야만 정상적으로 코드가 링크되는 것을 확인하였고, android 타겟 모듈 제외하고 전부를 kotlin 폴더 하위에 있도록 다시 지정하였습니다. 그리고 정상 빌드가 안되다 보니 PR을 분리하지는 않았습니다. [https://github.com/IRACK000/FilledSliderCompose/commit/bebad2003961f18af88e0d8aff3b2de13abeee37](url) 커밋 참고하시면 되겠습니다.

5. settings.gradle.kts에 jitpack.io는 Deprecated된 Repository여서 해당 부분 필수적이지 않다면 제거하는 것이 좋을 것 같습니다. 다른 repository에 대한 것도 Compose Multiplatform을 추가하는데 필수적이지 않은 경우 제거하는 것이 좋을 것 같습니다.

5A. 제가 알기로는 deprecated된것은 jcenter이지 jitpack이 아닌 것으로 알고 있습니다. 다만 역시 필수적이지 않기 때문에 새롭게 PR 올리면서 추가하지 않았습니다.

6. minSDK가 26에서 24로 바뀌었는데, 일반적으로 처음 minSDK로 정한 값보다 낮은 값으로 내리는 경우는 불가피한 경우가 아니면 권장되지 않습니다. 특히 안드로이드의 경우 minSDK 26부터 쓸 수 있는 API들이 많고, Compose를 사용하는 애플리케이션은 주로 minSDK가 26이상인 경우가 많기 때문에 오픈소스 개발 시에는 minSDK를 26으로 유지하는 것이 좋을 것으로 보입니다.

6A. minSDK 또한 기본 멀티플랫폼 템플릿이 24로 되어있어서 별 생각없이 거기에다 적용한거라 사실 PR 리뷰하시면서 수정하시겠지 하고 있던 부분이었습니다. 새 PR에서는 애초에 건드리지 않았습니다.


사실 이정도까지 수정하는데 오래 걸릴거라고 예상을 못하고 있었습니다.ㅜㅜ 안해본 프로젝트 구조를 건드리다 보니까 이렇게 되었네요.

나중에 시간 있으실 때 천천히 확인 해주시고, 좋은 라이브러리 개발해주셔서 감사하다는 말씀 드리고 싶습니다.
감사합니다!

b-re-w commented:
안녕하세요! 리뷰해주신 대로 수정하느라 시간이 좀 걸려서 답변이 늦었습니다.

`
아래는 리뷰주신 내용에 대한 반영내역입니다.
- 일단 전반적으로 제가 라이브러리를 멀티플랫폼화 시키면서 기존의 빌드 구조를 바탕으로 작업하기 보다는 Compose Multiplatform 공식 템플릿에 기존 코드를 옮기는 방향으로 작업했기 때문에 변경의 영향이 너무 넓었던 것 같습니다. 제가 PR 받는 쪽 입장을 제대로 고려를 못했던 부분인것 같고, 그래서 변경 내역을 전부 지우고 다시 기존 프로젝트 구조를 기반으로 변경 내역을 하나씩 도입하였습니다. PR도 최대한 나눠보았습니다.

`
1R) Pull Request의 단위가 분할 되어야 할 것 같습니다. 현재 올려주신 Pull Request에는 Gradle 업데이트, 모듈명 변경(sample-app --> android, main --> commonMain), target 버전 업데이트, 버전 관리 시 toml 적용 등이 한 번에 들어가 있는데, 한 번에 머지되어야 하는 코드 크기가 너무 커서 머지되면 영향 범위가 너무 넓습니다.
*만약 compose multiplatform을 지원하기 위해 target 버전 업데이트나 toml 적용이 필요가 없다면, Pull Request에서는 빼는 것이 좋을 것 같습니다. 만약 필요하다면, 사용하는 버전의 업데이트에 대한 PR이 먼저 올라가고, Compose Multiplatform에 대한 PR이 올라가야 할 것 같습니다.

1A) 말씀하신 부분 반영하기 위해서 독립적으로 실행 가능한 변경사항들끼리 최대한 분리하여 PR 만들었습니다. 모듈명과 프로젝트 구조도 기존 구조에서 변경하지 않기 위해서 최대한 노력했습니다. 그리고 필수가 아니었던 버전 관리 toml도 애초에 프로젝트 템플릿을 사용하지 않았기 때문에 굳이 적용하지 않았습니다. 타겟 버전 업데이트의 경우 멀티플랫폼을 지원하기 위해 코틀린 버전을 1.9.0으로 업데이트하고, 그에 따른 컴포즈 버전도 1.5.0으로 업데이트를 할 수밖에 없었습니다...(버전 안올리니까 오류가 나서 어쩔 수 없었습니다.) 버전 올리는 것에 대한 영향 검토해주시면 될거같습니다. #2 

2R) Compose Multiplatform을 위해 모듈명을 바꿔야 한다면, sample-app --> android로 바꾸기 보다, 더욱 명확하게 sample-app-android, sample-app-desktop 로 바꾸는 것이 좋을 것 같습니다. 또한 모듈명에 common이 들어가는 경우 공통으로 사용되지만 어떤 경우에 사용되어야 하는지 모호한 경우가 많기 때문에 기능이 들어간 main 모듈은 그대로 이름을 두는 것이 좋을 것 같습니다.

2A) 타겟별 sample app 모듈명을 바꾸지 않으려 노력을 좀 많이 했습니다. 제가 원래 올렸던 PR의 프로젝트 구조는 컴포즈 멀티플랫폼 기본 템플릿이라 그걸 그대로 사용하면 되었었는데, 타겟별로 분리하지 않고 한 폴더 안에 다 모아두려 하니 변경해야 할 부분들이 많았습니다. 샘플 앱 관련 PR은 전부 하나의 PR #4 에 모아두었습니다. 루트에 있던 common 모듈의 경우 CMP에서는 일반적으로 common에다 플랫폼 독립적인 코드를 작성하고, 각 타겟별 모듈(android, desktop, ios)에 플랫폼 종속적인 코드를 작성하는 방식으로 하다보니 common이 따로 나와있었는데, 아무래도 이 레포가 앱이 아닌 라이브러리에 대한 레포다 보니, 레포 루트에 android, desktop, ios, common이 다 나와있으면 정리 안된 느낌이 들거 같긴 해서 기존 sample-app폴더 안에 전부 정리해서 넣어두었습니다. common을 없애기 위해 sample-app 폴더가 common 역할을 하도록 했고, 타겟별 모듈은 sample-app 폴더 안에 target 폴더를 두어 그 안에 배치해두었습니다.
main모듈 이름을 그대로 두는 것이 좋겠다고 말씀하셔서 그렇게 해보려고 시도를 해봤습니다. main -> commonMain 이런 형태로 바꾼 것은 제가 그렇게 한게 아니라 멀티플랫폼 플러그인이 그렇게 되어 있어서 그런거였는데, source directory 폴더를 기존 안드로이드 모듈명으로 지정;kotlin.srcDirs("src/main")하여 빌드를 해보니, 안드로이드 타겟 빌드에서만 정상 작동하고, 나머지 타겟에서는 소스 폴더를 인식하지 못해서 다시 되돌려놓았습니다. [https://github.com/IRACK000/FilledSliderCompose/commit/068aaacbcc23dd0dbf19b81161a3287ce34c692c](url) 커밋 참고하시면 되겠습니다.
* [https://kmp.jetbrains.com/](url) 여기를 참고하시면, fleet 에디터를 기반으로 한 새로운 멀티플랫폼용 템플릿 프로젝트 생성기를 확인하실 수 있으며, 이 프로젝트 구조도 common을 사용하지 않으며 하나의 모듈에서 모든 타겟을 (ios를 제외한) 대응하는 것으로 보입니다. 다만 제가 아직 여기에 익숙하지가 않고, 타겟중 ios만 빠져나와 있는데다 fleet을 사용해야 원활하게 사용할 수 있는 듯 하여 사용하지 않았습니다. 나중에 한번 참고하시면 될거같습니다.

3R) 샘플 앱에서 ComposeSliderTheme 같이 함수의 선언부가 변경된 경우(Default 값이 없어진 경우)가 보이는데, 선언이 변경된 이유를 설명해주시면 좋을 것 같습니다.(프로덕션 코드가 아닌 샘플 코드라 중요한 부분은 아닙니다.)

3A) 이 부분은 코드를 다 안보셔서 그런것 같습니다. ComposeSliderTheme의 경우 Platform 종속적인 코드가 포함되어 있습니다. 안드로이드 12부터 지원되는 dynamicColor를 지원하기 위해 안드로이드 버전을 체크하기 위한 Build클래스와 최종 결정된 컬러를 시스펨바에 반영하기 위해 호출되는 LocalView.current가 다른 플랫폼에서는 지원되지 않기 때문에 commonMain 내에서는 컴파일 될 수 없습니다. 따라서 actual과 expect 키워드를 사용해서 플랫폼별 코드를 분리해서 서로 다르게 ComposeSliderTheme를 컴파일 하도록 작성해두었었는데, default 값은 expect function에서 선언 가능하고 actual쪽에서는 사용하지 않기 때문에 빠져있는 것으로, 아무래도 androidMain쪽 코드만 보셨던 것 같습니다. 그래도 ComposeSliderTheme 자체가 expect-actual로 선언되는 것보다는 그 내부 코드들이 분리되는 편이 좋을 것 같아 새 PR에서는 컬러 선택부와 글꼴 선택부만 따로 분리하였습니다. [https://github.com/IRACK000/FilledSliderCompose/commit/7de8edce1cb5ac6f950c8e1ddd61484d4aee713d](url)

4R) 1번과 같은 이야기 인데, Source Directory(java to kotlin)가 바뀌어야 하는 것이 PR에 필수적이지 않은 경우 별도의 PR로 만들어지면 좋을 것 같습니다.

4A) java to kotlin으로의 이름 변경이 저도 처음에는 필수가 아니라고 생각하고 있었어서 java로 그대로 놔둬보려고 시도해봤습니다. 테스트 결과 안드로이드 타겟 빌드시에는 모든 코드가 java 폴더 아래에 위치하고 있어도 빌드가 되지만, desktop과 ios 빌드시에는 kotlin 폴더 아래에 위치하고 있어야만 정상적으로 코드가 링크되는 것을 확인하였고, android 타겟 모듈 제외하고 전부를 kotlin 폴더 하위에 있도록 다시 지정하였습니다. 그리고 정상 빌드가 안되다 보니 PR을 분리하지는 않았습니다. [https://github.com/IRACK000/FilledSliderCompose/commit/bebad2003961f18af88e0d8aff3b2de13abeee37](url) 커밋 참고하시면 되겠습니다.

5R) settings.gradle.kts에 jitpack.io는 Deprecated된 Repository여서 해당 부분 필수적이지 않다면 제거하는 것이 좋을 것 같습니다. 다른 repository에 대한 것도 Compose Multiplatform을 추가하는데 필수적이지 않은 경우 제거하는 것이 좋을 것 같습니다.

5A) 제가 알기로는 deprecated된것은 jcenter이지 jitpack이 아닌 것으로 알고 있습니다. 다만 역시 필수적이지 않기 때문에 새롭게 PR 올리면서 추가하지 않았습니다.

6R) minSDK가 26에서 24로 바뀌었는데, 일반적으로 처음 minSDK로 정한 값보다 낮은 값으로 내리는 경우는 불가피한 경우가 아니면 권장되지 않습니다. 특히 안드로이드의 경우 minSDK 26부터 쓸 수 있는 API들이 많고, Compose를 사용하는 애플리케이션은 주로 minSDK가 26이상인 경우가 많기 때문에 오픈소스 개발 시에는 minSDK를 26으로 유지하는 것이 좋을 것으로 보입니다.

6A) minSDK 또한 기본 멀티플랫폼 템플릿이 24로 되어있어서 별 생각없이 거기에다 적용한거라 사실 PR 리뷰하시면서 수정하시겠지 하고 있던 부분이었습니다. 새 PR에서는 애초에 건드리지 않았습니다.

`
사실 이정도까지 수정하는데 오래 걸릴거라고 예상을 못하고 있었습니다.ㅜㅜ 안해본 프로젝트 구조를 건드리다 보니까 이렇게 되었네요.

나중에 시간 있으실 때 천천히 확인 해주시고, 좋은 라이브러리 개발해주셔서 감사하다는 말씀 드리고 싶습니다.
감사합니다!
